amida-tech/orange-api
# Feature/orange 201 update mongo

# what does this do
This updates several libraries and the use of those libraries to support MongoDB 3.6 (and probably 4.0 !  though I havent tested that...)
The two main areas that required change (besides the dependency versions) were as follows:
1. several places in the tests were creating subdocuments on patients in a way that caused `save` to be called on the patient twice in parallel - which spit out warnings and caused inconsistent errors
    * to address, I just rewrote these sections to perform the steps sequentially using promises
2. the gridfs-stream library which we were using for uploading/downloading avatars to/from mongodb GridFS doesn't support mongo3.6. This actually makes sense because the latest node driver for mongo has the same functionality built in. 
    * To address, I replaced the calls to the gridfs-stream library with the new calls to the updated mongo driver.

Also included in this PR is a small change to .env.example. I uncommented the mongo uri. This allows the travis CI tests to run again!

There are two failing tests. One, related to dose creation, seems to have slipped in while travis wasn't doing its job. The other fails sometimes - it is the geolocation test which uses a throttled external API (when you call it too many times, it fails).

# how to test
run `npm install`
stop whatever mongo you're usng and run mongo3.6 instead. I recommend using docker like so:

```
docker run --rm --name amida-orange-api-db -p 27017:27017 mongo:3.6
```

run `grunt test` and verify all but the two tests mentioned above pass

use the orange app and make sure things still work

# Also
`mongoose-unique-validator` will need to be updated when a new version gets published. They are calling a deprecated method in mongoose. They've since fixed it, but havent published a new version with the fix yet.
